replace variables and ~ in script paths

bebehei opened the pull request:
wordexp replaces all variables and `~` in the string. TBH I don't quite like the thing, that it replaces all variables, but there is no other way.

Fixes #350

bebehei commented:
Well, it might be good here, to add tests, too.

tsipinakis commented:
Yeah, tests are a good idea. `option_parser.c` already has tests for most of the functions so it should be trivial to extend them for the path one.

bebehei commented:
> Well, it might be good here, to add tests, too.

Here they are.

bebehei commented on src/option_parser.c:
```diff
@@ -109,6 +110,19 @@ char *get_value(char *section, char *key)
+
+        wordexp_t replaced_path;
+        if (path && 0 == wordexp(path, &replaced_path, WRDE_NOCMD)){
+                path = g_strdup(replaced_path.we_wordv[0]);
```
@tsipinakis I'm missing here a `free(path);`, right?

tsipinakis replied:
Yeah you are missing a `free`, `path` from `init_get_string` is never freed.

tsipinakis replied:
You're also only getting the first word here, we_wordv is a string array (`char **`), not a `char *`. It's a common mistake so please also add a regression test for this.

tsipinakis commented on test/option_parser.c:
```diff
@@ -59,6 +61,39 @@ TEST test_ini_get_string(void)
+        free(ptr);
+
+        path = ini_get_string(section, "expand_home", "");
+        wordexp(path, &replaced_path, WRDE_NOCMD);
```
Since we know the contents of the expansion we should use the simplest method (read: `getenv`) to get the data we need to compile the test case. Using the exact same code the function you're testing is using is error prone and can miss quite a few bugs.

 (hint: You can use the `string_replace` function from `utils.c` to do the actual replacement)

tsipinakis commented on test/option_parser.c:
```diff
@@ -59,6 +61,39 @@ TEST test_ini_get_string(void)
+
+        path = ini_get_string(section, "expand_tilde", "");
+        wordexp(path, &replaced_path, WRDE_NOCMD);
+        ASSERT_STR_EQ(replaced_path.we_wordv[0], (ptr = ini_get_path(section, "expand_tilde", "")));
```
Same with this test, we should use the simplest method to setup the test case.

tsipinakis commented on src/option_parser.c:
```diff
@@ -109,6 +110,19 @@ char *get_value(char *section, char *key)
+        char *path = ini_get_string(section, key, def);
+
+        wordexp_t replaced_path;
+        if (path && 0 == wordexp(path, &replaced_path, WRDE_NOCMD)){
```
Some more error handling here would be nice, especially for `WRDE_BADVAL` and `WRDE_CMDSUM`.

bebehei commented:
*Disclaimer: I'm currently not at my laptop for the next few days and can't test it. I read the wordexp manpage over and over again until I fully understood the Problems.*

I have to conclude, that this is the shittiest approach possible.

Here is why:

> You're also only getting the first word here, we_wordv is a string array

Yes this is correct. So if there are multiple words, we have to strcat all `we_wordv` elements together with `$IFS`.

> [Field splitting is done using the environment variable $IFS. If it is not set, the field separators are space, tab and newline.](https://linux.die.net/man/3/wordexp)

Well, if someone sets `IFS=/', the returned path is broken.

Even worse: IFS is unset, the path contains spaces. So how should we deterministically chain the words again together?

Wordexp also does more stuff, I personally dislike. It also replaces other variables. IMO in our case only `$HOME` (maybe additionally `$USER`) should get expanded and nothing more.

> (hint: You can use the string_replace function from utils.c to do the actual replacement)

What about using this in the actual `ini_get_string`? I think about replacing the first two characters if they match `~/` with `getenv('HOME')`.

bebehei commented:
>> Field splitting is done using the environment variable $IFS. If it is not set, the field separators are space, tab and newline.
>
> Well, if someone sets `IFS=/', the returned path is broken.
>
> Even worse: IFS is unset, the path contains spaces. So how should we deterministically chain the words again together?

Well, it turns out to be true in all cases. `wordexp` makes classical word. This is uesless.

> What about using this in the actual `ini_get_string`? I think about replacing the first two characters if they match `~/` with `getenv('HOME')`.

So, I went for that. Here are the commits. Sorry for the spam on travis. In my previous commits, I failed using `strcat`. :facepalm:

tsipinakis commented:
> So, I went for that. Here are the commits. Sorry for the spam on travis. In my previous commits, I failed using strcat.

No worries, you're not a C programmer until you've wasted a few hours debugging weird memory issues :p 

It might be beneficial to also apply this to the `dmenu` and `browser` settings, they are also paths after all.

bebehei commented:
> It might be beneficial to also apply this to the dmenu and browser settings, they are also paths after all.

Thanks for the hint.

bebehei commented:
> It might be beneficial to also apply this to the dmenu and browser settings, they are also paths after all.

I moved the logics now to a function in `utils.h`, added cmdline support and dmenu and the browser use these settings now, too.

The tests from `test_ini_get_path` about correct expansion have moved now to a separate test in `utils.h`, while in `test_(ini|cmdline|option)_get_path` now only test, if expansion works. and the handling of cmdline precedence.

tsipinakis commented:
LGTM 👍 

Thank you for implementing this.
